ably/export_test.go

```diff
@@ -149,3 +149,11 @@ func (c *RESTChannels) Len() int {
 	defer c.mu.RUnlock()
 	return len(c.cache)
 }
+
+func (c *RealtimeChannel) SetState(state ChannelState, err error) error {
+	return c.setState(state, err)
+}
+
+func (ablyConn *Connection) SetState(state ConnectionState, err error, retryIn time.Duration) error {
```

Comment:
The problem with setting just the state in tests is that it can easily lead to inconsistent overall states that can't happen in real conditions, so the tests can't be trusted. For instance, here you're resetting the state to INITIALIZED, which doesn't happen otherwise, and thus the several variables that realtime instances carry around could be in a state that's incompatible with such state (e. g. having a `connection.key` set), maybe causing unexpected behavior.

We should instead transition the state machine the via external events, as it's supposed to be the case. We have [ablytest.TransitionConn](https://github.com/ably/ably-go/blob/841b4e774670b72f648aa889c22b9f3e6b9cb261/ably/transitioner_test.go#L33) to make this explicit and deterministic.

Reply:
This is really helpful 🙏. Seems I am getting some race conditions because of incompatible states across the client. I will go through the above recommended solution.

Reply:
Resolved, we don't have `setState` anymore in the code